Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean

```diff
@@ -0,0 +1,648 @@
+  and `lTensor.equiv` gives a LinearEquiv.
+
+* For `N : Submodule R M`, `LinearMap.exact_subtype_mkQ N` says that
+  the inclusion of the submodule and the quotient map form an exact pair,
+  and `lTensor_mkQ` compute `ker (lTensor Q (N.mkQ))` and similarly for `rTensor_mkQ`
+
+* `TensorProduct.map_ker` computes the kernel of `TensorProduct.map f g'`
+  in the presence of two short exact sequences.
+
+The proofs are those of [bourbaki1989] (chap. 2, §3, n°6)
+
+* Analogue for morphisms of algebras :
+  `Algebra.TensorProduct.ker_map` computes the kernel of `Algebra.TensorProduct.map f g`
+
+* The two more specific lemmas `Algebra.TensorProduct.lTensor_ker`
+  and `Algebra.TensorProduct.rTensor_ker` compute the kernels
+  of `Algebra.TensorProduct.map f id` and `Algebra.TensorProduct.map id g`
+
+## TODO
+
+* Treat the noncommutative case
+
+* Treat the case of modules over semirings
+  (For a possible definition of an exact sequence of commutative semigroups, see
+  [Grillet-1969b], Pierre-Antoine Grillet,
+  *The tensor product of commutative semigroups*,
+  Trans. Amer. Math. Soc. 138 (1969), 281-293, doi:10.1090/S0002-9947-1969-0237688-1 .)
+
+-/
+
+section Exact
+
+variable {M N P : Type*} (f : M → N) (g : N → P)
+
+/-- The maps `f` and `g` form an exact pair :
+  `g y = 0` iff `y` belongs to the image of `f` -/
+def Exact [Zero P] : Prop := ∀ y, g y = 0 ↔ y ∈ Set.range f
+
+lemma Exact.comp_eq_zero [Zero P] (h : Exact f g) : g.comp f = 0 :=
+  funext fun _ => (h _).mpr <| Set.mem_range_self _
+
+open LinearMap
+
+variable {R : Type*} [CommSemiring R]
+variable {M N P : Type*}
+  [AddCommMonoid M] [AddCommMonoid N] [AddCommMonoid P]
+  [Module R M] [Module R N] [Module R P]
+
+variable {f : M →ₗ[R] N} {g : N →ₗ[R] P}
+
+lemma Exact.linearMap_ker_eq (hfg : Exact f g) : ker g = range f :=
+  SetLike.ext hfg
+
+lemma LinearMap.exact_iff : Exact f g ↔ LinearMap.ker g = LinearMap.range f :=
+  Iff.symm <| SetLike.ext_iff
+
+lemma Exact.linearMap_comp_eq_zero (h : Exact f g) : g.comp f = 0 :=
+  FunLike.coe_injective h.comp_eq_zero
+
+end Exact
+
+section Modules
+
+open TensorProduct LinearMap
+
+section Semiring
+
+variable {R : Type*} [CommSemiring R]
+variable {M N P P' : Type*}
+  [AddCommMonoid M] [AddCommMonoid N] [AddCommGroup P] [AddCommMonoid P']
+  [Module R M] [Module R N] [Module R P] [Module R P']
+
+variable {f : M →ₗ[R] N} {g : N →ₗ[R] P}
+
+variable {Q : Type*} [AddCommMonoid Q] [Module R Q]
+
+variable (g)
+
+lemma le_comap_range_lTensor (q : Q) :
+    LinearMap.range g ≤ (LinearMap.range (lTensor Q g)).comap (TensorProduct.mk R Q P q) := by
+  rintro x ⟨n, rfl⟩
+  exact ⟨q ⊗ₜ[R] n, rfl⟩
+
+lemma le_comap_range_rTensor (q : Q) :
+    LinearMap.range g ≤ (LinearMap.range (rTensor Q g)).comap
+      ((TensorProduct.mk R P Q).flip q) := by
+  rintro x ⟨n, rfl⟩
+  exact ⟨n ⊗ₜ[R] q, rfl⟩
+
+variable (Q) {g}
+
+/-- If `g` is surjective, then `rTensor Q g` is surjective -/
```

Comment:
Silly nitpick: should we try to always have the statement about `lTensor` first, and then the one for `rTensor`? (Or vice versa, I just picked the order that was used above for `le_comap_range_{l,r}Tensor`.)

Can you easily deduce one from the other, using commutativity of the tensor product? Or does that quickly become a nasty mess?

Reply:
It would be good practice to keep the same order throughout.
I will try to deduce `rTensor` from `lTensor`. This may be not so nice, though, if I need to require more module structures than the plain one. I'll see.

Reply:
Regarding your suggestion of deducing one from another : 
* I defined `TensorProduct.flip` and wrote few lemmas about it 
* That allows a quick proof for surjectivity for `lTensor` given that of `rTensor`  and (hopefully, to come) a quick proof of exactness for `lTensor` from that of `rTensor`.
However, the proof of exactness gives additional results that explicitely compute some quotients, and I wonder whether this shouldn't be retained. If so, it would be bizarre that they have different definitions, but maybe not. So I leave this open for the moment.

Reply:
What was the statement of your `TensorProduct.flip`?